staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go

```diff
@@ -162,6 +164,9 @@ type configController struct {
 	// name to the state for that level.  Every name referenced from a
 	// member of `flowSchemas` has an entry here.
 	priorityLevelStates map[string]*priorityLevelState
+
+	// totalLimitConsumers are to be kept appraised of the overall limits
+	totalLimitConsumers []func(maxWaitingRequests, maxExecutingRequests int)
```

Comment:
I think I'm still struggling with understand why we need to complicate the code so much for metrics.

What I would do is simply in lines 546-548 hardcode the metrics (which are P&F metrics anyway) and do something like:
```
fcmetrics.ReadWriteConcurrencyGaugeVec.NewForLabelValuesSafe(0, 1, []string{fcmetrics.LabelValueWaiting, epmetrics.ReadOnlyKind}),SetDenominator(meal.MaxWaiting)
fcmetrics.ReadWriteConcurrencyGaugeVec.NewForLabelValuesSafe(0, 1, []string{fcmetrics.LabelValueWaiting, epmetrics.MutatingKind}).SetDenominator(meal.MaxWaiting)
(and same for mutating)
```

I think that passing thing around, all those fields etc. is making this code very hard to follow...

Reply:
The problem is that lines 546--548 in apf_controller.go need to do what is primarily an up-call, to the higher layer in priority-and-fairness.go that updates the watermark data structures there and in maxinflight.go.  You have proposed a solution that folds in the knowledge of how those watermark data structures are composed, lessening the modularity.  That is not a deal-breaker, but something to consider.

I am struggling to remember why apf_filter.go:Interface allows registration of a multiplicity of request limit consumers; I think only one is needed.

Note that every call to NewForLabelValuesSafe returns a distinct object --- with its own denominator variable.

Reply:
The vector structure there is not really needed.   We statically know that it has exactly four members.  We can replace the vector with four individuals; that will eliminate the calls to NewForLabelValuesSafe with their attendant new object every time.

Reply:
No, wait, we need to use a vector so that all four time series show up as one metric with different label sets.

Reply:
I revised, using a combination of the not-incorrect ideas above.